### PR TITLE
[IOPID-2489] Add app signature verification before opening it

### PR DESCRIPTION
**Pull request by ChrisMattew**
## Short description
This PR adds app signature verification in the `isCieIdAvailable` and `launchCieIdForResult` methods before executing them.
## List of changes proposed in this pull request
- Added `isSignatureValid` function to verify if app can be opened
- Added native methods types declaration

## Demo
|Before       |After (fake `CieID` app)     |After (real `CieID` app)     |
|:----------|:----------------------------|:-----------------------|
|<video src="https://github.com/user-attachments/assets/6324dbf6-331d-4d7f-a481-563e785ed38b"></video>|<video src="https://github.com/user-attachments/assets/97da1c22-6b43-43e7-960d-15d62ff80ba8"></video>|<video src="https://github.com/user-attachments/assets/f28f7084-4434-40a7-a769-346bce0c436c"></video>|

>[!Tip]
>Tests have also been conducted on the `CieID` apps available in various alternative `Android` stores. Currently, among the known stores, it can be found on `APKPure` and `Aptoide`. 
## How to test
Using the sample app, replicate what is shown in the videos.

>[!Important]
>If you need the mock `CieID` app, I can privately provide the `APK` used in this test.

### CODE REVIEW

File: src/index.tsx [hunk 1]
```diff
@@ -9,13 +9,15 @@ const LINKING_ERROR =
...
-      {},
+      { isAppInstalled: () => false, launchCieIdForResult: () => {} },
       {
         get() {
           throw new Error(LINKING_ERROR);
         },
       }
     );
+const CIEID_SIGNATURE =
+  '92:D1:35:40:D4:50:F6:9F:79:2C:5F:3C:77:0A:E2:85:5B:FB:23:58:B4:47:A8:DE:06:4D:51:D0:35:8E:B6:97';
```
Comment (shadowsheep1) [inline src/index.tsx:20]: This package allow us to try a login with IPZS CieID for testing (`'it.ipzs.cieid.collaudo'`). If we force the signature verification also in UAT, we loose this possibility. Am I right? 

During my tests I've found that the CieID for testing show me this sha256: `6A:1C:79:CD:A2:61:C0:F9:73:60:9E:FE:3E:A4:B9:AB:C3:4D:C6:22:C4:2C:25:96:76:47:B1:A1:6E:24:67:C8`. We have two choices. Don't check the signature in UAT or ask IPZS if all their CieID App for testing present the above sha256 certificate digest. What do you think?

File: src/declarations.d.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    forwardUrl: string,
+    callback: (result: CieIdReturnType) => void
+  ) => void;
+}
+
+declare module 'react-native' {
+  interface NativeModulesStatic {
+    IoReactNativeCieidModule: IoReactNativeCieidModule;
+  }
+}
```
Comment (shadowsheep1) [inline src/declarations.d.ts]: Is this related to the topic of this PR? I guess not. Am I right? 
The proposition is right but this introduce a maintenance effort IMHO. What do you think? If we want to leve this changes what about adding some notes on our README file? What do you think?

File: src/index.tsx [hunk 3]
```diff
@@ -9,13 +9,15 @@ const LINKING_ERROR =
 const IoReactNativeCieidModule = NativeModules.IoReactNativeCieidModule
   ? NativeModules.IoReactNativeCieidModule
   : new Proxy(
-      {},
+      { isAppInstalled: () => false, launchCieIdForResult: () => {} },
```
Comment (shadowsheep1) [inline src/index.tsx]: As above

File: android/src/main/java/com/ioreactnativecieid/IoReactNativeCieidModule.kt [hunk 4]
```diff
@@ -28,9 +29,40 @@ class IoReactNativeCieidModule(reactContext: ReactApplicationContext) :
...
   @ReactMethod(isBlockingSynchronousMethod = true)
-  fun isAppInstalled(packageName: String): Boolean = try {
+  fun isAppInstalled(packageName: String, signature: String): Boolean = try {
     reactApplicationContext.packageManager.getPackageInfo(packageName, 0)
+    if(!isSignatureValid(packageName, signature)) {
+      throw PackageManager.NameNotFoundException()
+    }
     true
   } catch (e: PackageManager.NameNotFoundException) {
     false
```
Comment (shadowsheep1) [inline android/src/main/java/com/ioreactnativecieid/IoReactNativeCieidModule.kt]:
```suggestion
  @ReactMethod(isBlockingSynchronousMethod = true)
  fun isAppInstalled(packageName: String, signature: String): Boolean = try {
    reactApplicationContext.packageManager.getPackageInfo(packageName, 0)
    runCatching { isSignatureValid(packageName, signature) }.getOrDefault(false)
  } catch (e: PackageManager.NameNotFoundException) {
    false
```
What about this version?
- We don't tie the meaning of the `NameNotFoundException` exception to a signature mismatch.
- If for some reason the method isSignatureValid throws an exception different from `NameNotFoundException ` we don't crash the app.

What do you think?

File: android/src/main/java/com/ioreactnativecieid/IoReactNativeCieidModule.kt [hunk 5]
```diff
@@ -28,9 +29,40 @@ class IoReactNativeCieidModule(reactContext: ReactApplicationContext) :
...
+      val digest = MessageDigest.getInstance("SHA-256")
+      digest.update(it.toByteArray())
+      // Converts bytes in hexadecimal uppercase string
+      digest.digest().joinToString(":") { byte ->
+        "%02x".format(byte).uppercase()
+      }
+    }
+    // Returns the first sha256 signature
+    return sha256List.contains(signature)
+  }
```
Comment (shadowsheep1) [inline android/src/main/java/com/ioreactnativecieid/IoReactNativeCieidModule.kt]: What about leveraging extension function for more readability and maintenance? What do you think? 

```suggestion
  private fun isSignatureValid(packageName: String, signature: String): Boolean {
    val pm = reactApplicationContext.packageManager
    val packageInfo = pm.getPackageInfoCompat(packageName)

    val signatures = packageInfo.getSignaturesCompat()

    val sha256List = signatures.map { it.toSha256() }

    // Check if the given signature is in the SHA-256 list
    return sha256List.contains(signature)
  }

  // Extension function to handle API compatibility for getPackageInfo
  private fun PackageManager.getPackageInfoCompat(packageName: String): PackageInfo {
    return if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.P) {
      getPackageInfo(packageName, PackageManager.GET_SIGNING_CERTIFICATES)
    } else {
      @Suppress("DEPRECATION")
      getPackageInfo(packageName, PackageManager.GET_SIGNATURES)
    }
  }

  // Extension function to handle API compatibility for getting signatures
  private fun PackageInfo.getSignaturesCompat(): Array<Signature> {
    return if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.P) {
      signingInfo.apkContentsSigners
    } else {
      @Suppress("DEPRECATION")
      signatures
    }
  }

  // Extension function to convert a signature to its SHA-256 hash
  private fun Signature.toSha256(): String {
    val digest = MessageDigest.getInstance("SHA-256")
    return digest.digest(toByteArray())
      .joinToString(":") { byte -> "%02x".format(byte).uppercase() }
  }
```

File: android/src/main/java/com/ioreactnativecieid/IoReactNativeCieidModule.kt [hunk 6]
```diff
@@ -51,6 +84,11 @@ class IoReactNativeCieidModule(reactContext: ReactApplicationContext) :
       }
 
       try {
+        if(!isSignatureValid(packageName, signature)) {
+          onActivityResultCallback = null
+          ME.CIEID_SIGNATURE_MISMATCH.invoke(resultCallback)
+          return;
```
Comment (shadowsheep1) [inline android/src/main/java/com/ioreactnativecieid/IoReactNativeCieidModule.kt]:
```suggestion
          return
```
We don't need that "old" terminator :P

**Review by shadowsheep1 — CHANGES_REQUESTED** [review]
What about changing/integrating our README files (Main and Example One)? 
What about double check them to see what to remove (old stuff) and what to add now that we are adding more checks?
What do you think?

File: android/src/main/java/com/ioreactnativecieid/IoReactNativeCieidModule.kt (hunk 6 above)
Comment (ChrisMattew) [reply android/src/main/java/com/ioreactnativecieid/IoReactNativeCieidModule.kt]: Good to know! Thanks! 👀

File: src/index.tsx (hunk 1 above)
Comment (ChrisMattew) [reply src/index.tsx:20]: Good catch! 
In my opinion, there could be two viable scenarios:  
- **First scenario**: We ignore the signature verification on the test app in this initial implementation and introduce it later once we receive a response from IPZS.  
- **Second scenario**: We include the SHA256 hash you shared and use that to open the test app while, in the meantime, asking IPZS if the signature in question is correct.  

What do you think?

File: src/declarations.d.ts (hunk 2 above)
Comment (ChrisMattew) [reply src/declarations.d.ts]: Yes you are right! It's not related to the topic of this PR. Even if this methods are not exposed to the final user and they could be removed without affecting the UX of them, I think they could improve the maintainers' DX of this package.
However, I agree with you about adding some notes on our README file.

File: src/index.tsx (hunk 3 above)
Comment (ChrisMattew) [reply src/index.tsx]: As above 👀

File: android/src/main/java/com/ioreactnativecieid/IoReactNativeCieidModule.kt (hunk 4 above)
Comment (ChrisMattew) [reply android/src/main/java/com/ioreactnativecieid/IoReactNativeCieidModule.kt]: Great suggestion! I think the solution you suggest is more correct and more elegant than mine, thanks!

File: src/index.tsx (hunk 1 above)
Comment (shadowsheep1) [reply src/index.tsx:20]: Personally I like the first one more.

File: src/declarations.d.ts (hunk 2 above)
Comment (shadowsheep1) [reply src/declarations.d.ts]: If we change our functions signature we've to remember to change this part too. So we need a way that allows us not to forget it.

File: android/src/main/java/com/ioreactnativecieid/IoReactNativeCieidModule.kt (hunk 5 above)
Comment (ChrisMattew) [reply android/src/main/java/com/ioreactnativecieid/IoReactNativeCieidModule.kt]: I can only agree with you; thank you for the suggestion, I wasn't aware of this approach!

File: src/index.tsx (hunk 1 above)
Comment (ChrisMattew) [reply src/index.tsx:20]: Addressed [here](https://github.com/pagopa/io-react-native-cieid/pull/10/commits/94e75dd19bf7b38e6bf0cccd6758aa225df88fb5)

File: src/declarations.d.ts (hunk 2 above)
Comment (ChrisMattew) [reply src/declarations.d.ts]: This file has been removed!

File: android/src/main/java/com/ioreactnativecieid/IoReactNativeCieidModule.kt [hunk 7]
```diff
@@ -28,10 +29,47 @@ class IoReactNativeCieidModule(reactContext: ReactApplicationContext) :
...
+    return digest.digest(toByteArray())
+      .joinToString(":") { byte -> "%02x".format(byte).uppercase() }
+  }
+
   @ReactMethod(isBlockingSynchronousMethod = true)
-  fun isAppInstalled(packageName: String): Boolean = try {
+  fun isAppInstalled(packageName: String, signature: String? = null): Boolean = try {
     reactApplicationContext.packageManager.getPackageInfo(packageName, 0)
-    true
+    if(signature == null) {
```
Comment (shadowsheep1) [inline android/src/main/java/com/ioreactnativecieid/IoReactNativeCieidModule.kt]:
```suggestion
    if (signature == null) {
```

I see some formatting issues. Try to format the whole file with Android Studio. There are a few ways to do that, personally I use: `Command+Option+L`.

**Review by shadowsheep1 — CHANGES_REQUESTED** [review]

File: android/src/main/java/com/ioreactnativecieid/IoReactNativeCieidModule.kt (hunk 7 above)
Comment (ChrisMattew) [reply android/src/main/java/com/ioreactnativecieid/IoReactNativeCieidModule.kt]: Thanks for the shortcut! Addressed [here](https://github.com/pagopa/io-react-native-cieid/pull/10/commits/eac019a2570d2c99cde7bc9edeadb31290751bef)

**Review by shadowsheep1 — APPROVED** [review]
LGTM
